discovery_engine_core/core/src/stack/exploration.rs

```diff
@@ -0,0 +1,531 @@
+        let docs = new_docs_with_embeddings(&[
+            [3., 1., 0.],
+            [-1., 0., 0.],
+            [2., 3., 0.],
+            [-3., 7., 0.],
+            [-2., 5., 0.],
+        ]);
+
+        let expected = vec![docs[1].id, docs[3].id];
+
+        let pos_cois = create_pos_cois(&[[4., 4., 0.], [5., 5., 0.], [6., 4., 0.]]);
+        let neg_cois = create_neg_cois(&[[4., 1., 0.], [5., 1., 0.], [5., -1., 0.]]);
+
+        let config = Config {
+            number_of_candidates: 3,
+            ..Config::default()
+        };
+        let mut rng = ChaCha8Rng::seed_from_u64(1);
+        let docs =
+            document_selection_with_rng(&pos_cois, &neg_cois, docs, &config, &mut rng).unwrap();
+
+        assert_eq!(docs.len(), 2);
+        assert_eq!(docs[0].id, expected[0]);
+        assert_eq!(docs[1].id, expected[1]);
+    }
+
+    #[test]
+    fn test_document_selection_close_no_documents() {
+        let pos_cois = create_pos_cois(&[[4., 4., 0.]]);
+        let config = Config {
+            number_of_candidates: 10,
+            ..Config::default()
+        };
+        let docs = document_selection(&pos_cois, &[], vec![], &config).unwrap();
+        assert_eq!(docs.len(), 0);
+    }
+
+    #[test]
+    fn test_document_selection_close_no_candidates() {
+        let docs = new_docs_with_embeddings(&[[3., 1., 0.]]);
+        let pos_cois = create_pos_cois(&[[4., 4., 0.]]);
+        let config = Config {
+            number_of_candidates: 0,
+            ..Config::default()
+        };
+        let docs = document_selection(&pos_cois, &[], docs, &config).unwrap();
+        assert_eq!(docs.len(), 0);
+    }
+
+    #[test]
+    fn test_document_selection_close_all_documents() {
+        let docs =
+            new_docs_with_embeddings(&[[3., 1., 0.], [-1., 0., 0.], [2., 3., 0.], [-3., 7., 0.]]);
+        let pos_cois = create_pos_cois(&[[4., 4., 0.]]);
+        let config = Config {
+            number_of_candidates: 4,
+            ..Config::default()
+        };
+        let docs = document_selection(&pos_cois, &[], docs, &config).unwrap();
+        assert_eq!(docs.len(), 4);
+    }
+
+    #[test]
+    fn test_document_selection_close_no_cois() {
+        let res = document_selection(&[], &[], vec![], &Config::default());
+        assert!(matches!(res.unwrap_err(), Error::NotEnoughCois));
+    }
+
+    #[test]
+    fn test_document_selection_close_more_than_max() {
+        let docs = new_docs_with_embeddings(&[
+            [3., 3., 1.],
+            [-3., -3., 1.],
+            [-3., 3., 1.],
+            [3., -3., 1.],
+            [3., 0., 1.],
+            [-3., 0., 1.],
+            [0., 3., 1.],
+            [0., -3., 1.],
+        ]);
+        let expected = vec![docs[4].id, docs[5].id];
+        let pos_cois = create_pos_cois(&[[0., 0., 1.]]);
+        let config = Config {
+            number_of_candidates: 8,
+            max_selected_docs: 2,
+        };
+        let mut rng = ChaCha8Rng::seed_from_u64(1);
```

Comment:
:grey_question: This is to match the python tests, right?

Reply:
That was my first idea, but unfortunately there is no common rng between python and rust. But i also need a seeded rng so that the random selection for the doc stays the same between test runs. There is also `rand::rngs::StdRng` but i followed there recommendation in the docs:

> The algorithm is deterministic but should not be considered reproducible due to dependence on configuration and possible replacement in future library versions. For a secure reproducible generator, we recommend use of the rand_chacha crate directly.